nixops/nix_expr.py

```diff
@@ -156,13 +177,29 @@ def py2nix(value, initial_indentation=0, maxwidth=80, inline=False):
     True, squash everything into a single line.
     """
 
-    def _enc_int(node):
+    def _enc_int(node: int) -> RawValue:
         if node < 0:
             return RawValue("builtins.sub 0 " + str(-node))
         else:
             return RawValue(str(node))
 
-    def _enc_str(node, for_attribute=False):
+    @overload
+    def _enc_str(node: str, for_attribute: Literal[True]) -> str:
+        ...
+
+    @overload
+    def _enc_str(
+        node: str, for_attribute: Literal[False]
+    ) -> Union[RawValue, Container]:
+        ...
+
+    @overload
+    def _enc_str(node: str) -> Union[RawValue, Container]:
+        ...
+
+    def _enc_str(
+        node: str, for_attribute: bool = False
+    ) -> Union[str, RawValue, Container]:
```

Comment:
I'm curious to try this overall PR without the `Literal` import, given that it's only used in two places. If that works, we can merge the rest and make this it's own PR and figure out if/how python does dependency management.

Reply:
Yea, I'd kinda prefer separate functions with it clear in the name what they do/return, rather than relying on a special parameter, but the pattern was present in a variety of places and I don't know if there are calls to this stuff (particularly the `logged_exec` calls/wrappers) outside of this code base, so was hesitant to change function names.

If we can refactor and do away with all the `Literal` junk without also adding a bunch of `cast`s, I think that'd be a win.